### PR TITLE
Closes #80 and #85 Speed up derive functions and remove `by_vars`

### LINKED ISSUE #80: Feature Request:  Speed up the computation of `derive_params_growth_xxx()` functions
### Feature Idea

The `left_join()` with a many-to-many relationship code block blows up really quickly and slows down code run time
Can be optimized by:

1. create a age-bin categorical variable within both the input dataset and metadata to get a unique-key to match
2. use library(fuzzyjoin) which would introduce a dependency

### Relevant Input

_No response_

### Relevant Output

_No response_

### Reproducible Example/Pseudo Code

_No response_

### PR DESCRIPTION
**Pull request by zdz2101**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiralpeds (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiralpeds site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the Reference page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Review by zdz2101 — COMMENTED** [review]
@Fanny-Gautier If you want to try rerunning the larger dataset through this branch/version of the package it should be **significantly** speedier now through this methodology, but this is still in the works/good head start on the speed up process

P.S. not completely ready, the rounding issue is "back" how to identify the appropriate bin after rounding

File: R/bin_utils.R [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+#'   A metadata dataset that is the same as that in `derive_params_growth_age()`/`derive_params_growth_height()`
+#'
+#' @param param Character vector
+#'
+#'   Accepted values: "AGE"/"HEIGHT_LENGTH"
+#'
+#' @family internal
+#' @keywords internal
+#' @return list of breaks and labels for age/height bins
+get_bins <- function(meta_criteria, param) {
```
Comment (zdz2101) [inline R/bin_utils.R:16]: @rossfarrugia @Fanny-Gautier this function grabs the metadata AGE/HEIGHT_LENGTH variable and finds the unique "bins" that created agnostic of bin width, e.g [0,1) [1,2) [2, 3) ... etc

File: R/bin_utils.R [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+#' @param param Parameter
+#'
+#'  Vector containing parameter that needs closest bin checking
+#'
+#' @param breaks Breaks created in `get_bins()`
+#'
+#' @param labels Labels created in `get_bins()`
+#'
+#' @return detects appropriate bin
+find_closest_bin <- function(param, breaks, labels) {
```
Comment (zdz2101) [inline R/bin_utils.R:40]: this function will help with the rounding say we have L/M/S values for 25 and 26, there will be two unique "bins" created [25, 26) and [26, 27) for values of 25.4 we want it to map to the [25, 26) bin and for 25.5 we want it to map to [26, 27) bin , this function achieves that, agnostic of bin size, e.g say we have bins [1, 2) [2, 3) [3, 10) [10, 11) and the value 9 should map to 10 since its closer to 10 but is technically in the [3, 10) bin it change the bin accordingly

File: R/bin_utils.R [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+#' Set Age/Height Bins
+#'
+#' Adds the unique age/height-bins to each dataframe to have a column for `by` in join.
+#'
+#' @family internal
+#' @keywords internal
+#' @return Character vector of unique age/height bins based on metadata
+set_bins <- function(param, breaks, labels) {
```
Comment (zdz2101) [inline R/bin_utils.R:85]: This is just a wrapper for the `map()` statement later in the code

File: R/derive_params_growth_age.R [hunk 4]
```diff
@@ -239,11 +239,15 @@ derive_params_growth_age <- function(dataset,
...
+  bins <- get_bins(meta_criteria, param = "AGE")
+
   # create a unified join naming convention, hard to figure out in by argument
-  dataset <- dataset %>%
+  dataset2 <- dataset %>%
     mutate(
       sex_join := {{ sex }},
-      ageu_join := {{ age_unit }}
+      ageu_join := {{ age_unit }},
+      age_bins := map({{ age }}, ~ set_bins(.x, breaks = bins$breaks, labels = bins$labels))
```
Comment (zdz2101) [inline R/derive_params_growth_age.R:250]: By creating a column `age_bins`

File: R/derive_params_growth_age.R [hunk 5]
```diff
@@ -262,25 +266,18 @@ derive_params_growth_age <- function(dataset,
       SD2pos = (M * (1 + 2 * L * S)^(1 / L)),
       SD3pos = (M * (1 + 3 * L * S)^(1 / L)),
       SD2neg = (M * (1 - 2 * L * S)^(1 / L)),
-      SD3neg = (M * (1 - 3 * L * S)^(1 / L))
+      SD3neg = (M * (1 - 3 * L * S)^(1 / L)),
+      age_bins = map(metadata_age, ~ set_bins(.x, breaks = bins$breaks, labels = bins$labels))
```
Comment (zdz2101) [inline R/derive_params_growth_age.R:270]: and here

File: R/derive_params_growth_age.R [hunk 6]
```diff
@@ -262,25 +266,18 @@ derive_params_growth_age <- function(dataset,
...
 
   # Merge the dataset that contains the vs records and filter the L/M/S that fit the appropriate age
-  added_records <- dataset %>%
+  added_records <- dataset2 %>%
     filter(!!enexpr(parameter)) %>%
     left_join(.,
       processed_md,
-      by = c("sex_join", "ageu_join"),
-      relationship = "many-to-many"
+      by = c("sex_join", "ageu_join", "age_bins")
```
Comment (zdz2101) [inline R/derive_params_growth_age.R:278]: we can use it as a unique join key to significantly reduce the run time of the appropriate join

**Review by zdz2101 — COMMENTED** [review]
and as you can see the tests should still run and pass exactly the same

File: R/derive_params_growth_age.R [hunk 7]
```diff
@@ -262,25 +266,18 @@ derive_params_growth_age <- function(dataset,
...
     )
 
   # Merge the dataset that contains the vs records and filter the L/M/S that fit the appropriate age
-  added_records <- dataset %>%
+  added_records <- dataset2 %>%
     filter(!!enexpr(parameter)) %>%
     left_join(.,
       processed_md,
-      by = c("sex_join", "ageu_join"),
-      relationship = "many-to-many"
```
Comment (zdz2101) [inline R/derive_params_growth_age.R:274]: The many to many join was creating m x n rows where m is denoted as number of records in the input dataset, and n as number of records in the metadata, then filtered after figuring out the match after the fact, that's why this was running as slow as it was, I didn't realize the scale of how quickly it would blow up

File: R/derive_params_growth_height.R [hunk 8]
```diff
@@ -328,11 +326,11 @@ derive_params_growth_height <- function(dataset,
...
-      select(-c(L, M, S, sex_join, heightu_join, meta_height, ht_diff, is_lowest))
+      select(-c(L, M, S, sex_join, heightu_join, meta_height))
   }
 
   dataset_final <- dataset_final %>%
-    select(-c(SD2pos, SD3pos, SD2neg, SD3neg))
+    select(-c(SD2pos, SD3pos, SD2neg, SD3neg, ht_bins))
 
   return(dataset_final)
 }
```
Comment (Fanny-Gautier) [inline R/derive_params_growth_height.R:336]: Could you please double check where `temp_z` and `temp_val` should be dropped ? they are still in the final dataset. Thanks

File: R/derive_params_growth_age.R (hunk 5 above)
Comment (Fanny-Gautier) [inline R/derive_params_growth_age.R:270]: `age_bins` should be defined as a global variable, I get the following message while running the checks "_derive_params_growth_age: no visible binding for global variable 'age_bins'_ "

File: R/derive_params_growth_height.R [hunk 9]
```diff
@@ -227,11 +227,14 @@ derive_params_growth_height <- function(dataset,
...
 
+  bins <- get_bins(meta_criteria, param = "HEIGHT_LENGTH")
+
   # create a unified join naming convention, hard to figure out in by argument
   dataset <- dataset %>%
     mutate(
       sex_join := {{ sex }},
-      heightu_join := {{ height_unit }}
+      heightu_join := {{ height_unit }},
+      ht_bins := map({{ height }}, ~ set_bins(.x, breaks = bins$breaks, labels = bins$labels))
```
Comment (Fanny-Gautier) [inline R/derive_params_growth_height.R:237]: `ht_bins` should be defined as a global variable, I get the following message while running the checks "_derive_params_growth_height: no visible binding for global variable 'ht_bins'_ "

File: R/bin_utils.R [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+#' @keywords internal
+#' @return list of breaks and labels for age/height bins
+get_bins <- function(meta_criteria, param) {
+  if (param == "AGE") {
+    unique_vals <- sort(unique(meta_criteria$AGE))
+  } else if (param == "HEIGHT_LENGTH") {
+    unique_vals <- sort(unique(meta_criteria$HEIGHT_LENGTH))
+  }
+  breaks <- c(unique_vals, Inf)
+  labels <- paste0(head(breaks, -1), ",", tail(breaks, -1))
```
Comment (Fanny-Gautier) [inline R/bin_utils.R:24]: I get the following messages while running the checks. Could you please consider them accordingly:
- get_bins: no visible global function definition for 'head'
- get_bins: no visible global function definition for 'tail'
- Undefined global functions or variables: age_bins head ht_bins tail

**Review by Fanny-Gautier — APPROVED** [review]
unit tests are still matching, and updates are running fine.

File: R/derive_params_growth_age.R
Comment (Fanny-Gautier) [inline R/derive_params_growth_age.R:1]: From line 288 (# create separate records objects as appropriate depending if user specific sds and/or pctl), it creates extra records even when there is no selected records in `added_records`. Could you please double check with David's data and patient 7123 why some `HDCSDS` and `HDCPCTL` records are created while there is no record meeting `VSTESTCD == "HDCIRC"`

**Review by Fanny-Gautier — CHANGES_REQUESTED** [review]
Findings with SDS and PCTL records created while no data available. Please double check `derive_params_growth_age()`

File: R/derive_params_growth_age.R
Comment (zdz2101) [reply R/derive_params_growth_age.R:1]: Thanks for pointing that out, I should have pushed a fix to address this, I'm rerunning david's data to check again too

File: tests/testthat/test-derive_params_growth_age.R [hunk 11]
```diff
@@ -335,7 +335,7 @@ test_that("derive_params_growth_age Test 6: Test out of bound ages", {
     )
   )
 
-  expected <- c(NA_real_, NA_real_)
+  expected <- numeric()
```
Comment (zdz2101) [inline tests/testthat/test-derive_params_growth_age.R:338]: @Minlei0201 @rossfarrugia Can't quite wrapy my head around whether or not we really should be creating empty rows for unmatched records due to the nature of how the metadata needs to be split anyway, like in theory if you're running weight for age via who you'll have unmatched records for cdc and vice versa. I was running into issues trying to recreate david's dataset with the retention of unmatched records

**Comment by zdz2101** [comment]
@Fanny-Gautier @rossfarrugia it also occured to me with the way I programmed this speeding up of the function I dont think we need a `by_vars` argument anymore, perhaps we should remove it? Originally I needed it to match the appropriate metadata by searching for the best match per subject but now I have different functions are doing it in a vectorized way

File: tests/testthat/test-derive_params_growth_age.R (hunk 11 above)
Comment (Fanny-Gautier) [reply tests/testthat/test-derive_params_growth_age.R:338]: What are the pro and cons to create empty records? I don't think we need them, at least in the final dataset.
I had to manipulate David's dataset in order to compare the various results, I don't think that his dataset is in a proper shape, but we could imagine users having different metadata and resulting to such dataset.

File: R/derive_params_growth_age.R [hunk 12]
```diff
@@ -239,11 +239,16 @@ derive_params_growth_age <- function(dataset,
     cli_abort("One of `set_values_to_sds`/`set_values_to_pctl` has to be specified.")
   }
 
+  # Get the breaks and labels
+  bins <- get_bins(meta_criteria, param = "AGE")
+
   # create a unified join naming convention, hard to figure out in by argument
-  dataset <- dataset %>%
+  dataset2 <- dataset %>%
```
Comment (rossfarrugia) [inline R/derive_params_growth_age.R:246]: can we use more descriptive naming than `dataset2`?

File: R/derive_params_growth_age.R [hunk 13]
```diff
@@ -262,29 +267,22 @@ derive_params_growth_age <- function(dataset,
...
-    filter(is_lowest & row_number() == 1) %>%
-    ungroup()
-
+    filter(!is.na(metadata_age))
 
   by_exprs <- enexpr(by_vars)
   by_antijoin <- setNames(as.character(by_exprs), as.character(by_exprs))
-  unmatched_records <- anti_join(dataset, added_records, by = by_antijoin)
+  # unmatched_records <- anti_join(dataset2, added_records, by = by_antijoin) %>%
+  #   filter(!!enexpr(parameter))
```
Comment (rossfarrugia) [inline R/derive_params_growth_age.R:285]: can we remove any commented code blocks?

File: R/derive_params_growth_age.R [hunk 14]
```diff
@@ -324,11 +322,11 @@ derive_params_growth_age <- function(dataset,
...
     }
 
-    unmatched_sds <- unmatched_records %>%
-      mutate(!!!set_values_to_sds)
+    # unmatched_sds <- unmatched_records %>%
+    #   mutate(!!!set_values_to_sds)
 
-    dataset_final <- bind_rows(dataset, add_sds, unmatched_sds) %>%
-      select(-c(L, M, S, sex_join, ageu_join, metadata_age, age_diff, is_lowest, temp_val, temp_z))
+    dataset_final <- bind_rows(dataset, add_sds) %>% # , unmatched_sds) %>%
```
Comment (rossfarrugia) [inline R/derive_params_growth_age.R:328]: as above remove commented code

File: vignettes/advs.Rmd [hunk 15]
```diff
@@ -54,6 +54,7 @@ library(admiral)
 library(admiralpeds)
 library(dplyr, warn.conflicts = FALSE)
 library(lubridate)
+library(rlang)
```
Comment (rossfarrugia) [inline vignettes/advs.Rmd:57]: same needed for ADVS template?

**Review by rossfarrugia — COMMENTED** [review]
@zdz2101 just catching up and here's some minor review comments as well as the lintr CI fail needs fixing, but i trust yours and @Fanny-Gautier's call on this one if you think it is worth merging before we submit to CRAN if it optimizes code and we match with David's results.

**Comment by Fanny-Gautier** [comment]
@zdz2101 Could you please review the `BMI-for-age percentile` calculation ? It currently gives me the same results as the `BMI-for-age Z-score`. e.g. patient "17261" from `df_test_data_2.csv`
image
Thanks

**Comment by zdz2101** [comment]
> @zdz2101 Could you please review the `BMI-for-age percentile` calculation ? It currently gives me the same results as the `BMI-for-age Z-score`. e.g. patient "17261" from `df_test_data_2.csv` image Thanks

This should be fixed now

File: vignettes/advs.Rmd (hunk 15 above)
Comment (rossfarrugia) [reply vignettes/advs.Rmd:52]: ok that's odd. leave it then and sure nobody would even notice

**Review by rossfarrugia — APPROVED** [review]
All of my comments are resolved but I'd wait for @Fanny-Gautier to double check David's test data and approve before merging. I'll also share in the dev slack in case others have chance to look today.

**Comment by Fanny-Gautier** [comment]
> All of my comments are resolved but I'd wait for @Fanny-Gautier to double check David's test data and approve before merging. I'll also share in the dev slack in case others have chance to look today.

I cannot see the resolution for `BMI-for-age percentile` while I pulled the branch and have re-run everything. Maybe I am missing something.
image
image

File: R/derive_params_growth_age.R [hunk 16]
```diff
@@ -366,18 +344,15 @@ derive_params_growth_age <- function(dataset,
         select(-c(P95, Sigma))
     } else {
       add_pctl <- add_pctl %>%
-        mutate(AVAL = pnorm(AVAL) * 100)
+        mutate(AVAL = pnorm(temp_z) * 100)
```
Comment (zdz2101) [inline R/derive_params_growth_age.R:347]: @Fanny-Gautier I thought it was this piece that did the trick but let me double check again

**Review by Fanny-Gautier — APPROVED** [review]
Results matching my checks and David's results ! Approved !

File: R/derive_params_growth_age.R [hunk 17]
```diff
@@ -183,25 +177,22 @@
...
 #'   age = AGECUR,
 #'   age_unit = AGECURU,
 #'   meta_criteria = criteria,
-#'   parameter = VSTESTCD == "WEIGHT",
+#'   parameter = VSTESTCD == "HEIGHT",
 #'   analysis_var = VSSTRESN,
-#'   who_correction = TRUE,
 #'   set_values_to_sds = exprs(
-#'     PARAMCD = "WGASDS",
+#'     PARAMCD = "HTSDS",
```
Comment (Fanny-Gautier) [inline R/derive_params_growth_age.R:187]: PARAMCD should = "HGTSDS"

File: R/derive_params_growth_age.R [hunk 18]
```diff
@@ -183,25 +177,22 @@
...
 #'     PARAM = "Weight-for-age z-score"
 #'   ),
 #'   set_values_to_pctl = exprs(
-#'     PARAMCD = "WGAPCTL",
+#'     PARAMCD = "HTPCTL",
```
Comment (Fanny-Gautier) [inline R/derive_params_growth_age.R:191]: PARAMCD should = "HGTPCTL"

File: R/derive_params_growth_age.R [hunk 19]
```diff
@@ -183,25 +177,22 @@
...
 #'   age_unit = AGECURU,
 #'   meta_criteria = criteria,
-#'   parameter = VSTESTCD == "WEIGHT",
+#'   parameter = VSTESTCD == "HEIGHT",
 #'   analysis_var = VSSTRESN,
-#'   who_correction = TRUE,
 #'   set_values_to_sds = exprs(
-#'     PARAMCD = "WGASDS",
+#'     PARAMCD = "HTSDS",
 #'     PARAM = "Weight-for-age z-score"
```
Comment (Fanny-Gautier) [inline R/derive_params_growth_age.R:188]: Using `VSTESTCD == "HEIGHT"`, should we not have PARAM = "**Height**-for-age z-score" ?

File: R/derive_params_growth_age.R [hunk 20]
```diff
@@ -183,25 +177,22 @@
...
 #'   ),
 #'   set_values_to_pctl = exprs(
-#'     PARAMCD = "WGAPCTL",
+#'     PARAMCD = "HTPCTL",
 #'     PARAM = "Weight-for-age percentile"
```
Comment (Fanny-Gautier) [inline R/derive_params_growth_age.R:192]: Using `VSTESTCD == "HEIGHT"`, should we not have PARAM = "**Height**-for-age percentile" ?

**Review by Fanny-Gautier — CHANGES_REQUESTED** [review]
@zdz2101, Please review the example and correct as necessary

**Review by Fanny-Gautier — APPROVED** [review]
Last comments properly implemented. Ready to merge.
